src/portal-service.vala

```diff
@@ -837,6 +841,14 @@ namespace Frida {
 
 		private void on_agent_session_expired (AgentSessionEntry entry) {
 			sessions.unset (entry.id);
+
+			ClusterNode? node = entry.node;
+			if (node != null)
+				node.sessions.remove (entry.id);
+
+			ControlChannel? controller = entry.controller;
+			if (controller != null)
+				controller.sessions.remove (entry.id);
```

Comment:
Not super sure this is needed, though.

Reply:
Ah yes indeed, since the timer gets started by `detach_node_and_controller ()` and `detach_controller ()`, and in the former case `teardown_cluster_node ()` does this, and in the latter case we are tearing down the `ControlChannel`.